### PR TITLE
Balder/add in memory crc for summary cache

### PR DESCRIPTION
**Pull request by baldersheim**
@geirst and @toregge PR
Best reviewed in sequence. Only the last adds any semantics

### CODE REVIEW

**Review by toregge — APPROVED** [review]
LGTM

**Comment by toregge** [comment]
Please update unit tests, so they pass too.

File: searchlib/src/vespa/searchlib/docstore/value.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/value.cpp]: Please update year

File: searchlib/src/vespa/searchlib/docstore/value.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/value.h]: Please update year

File: searchlib/src/vespa/searchlib/docstore/value.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+#pragma once
+
+#include <vespa/vespalib/util/compressionconfig.h>
+#include <vespa/vespalib/data/databuffer.h>
+
+namespace search::docstore {
+
+class Value {
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/value.h]: Consider adding class comment

File: searchlib/src/vespa/searchlib/docstore/documentstore.cpp [hunk 4]
```diff
@@ -168,21 +172,32 @@ DocumentStore::visit(const LidVector & lids, const DocumentTypeRepo &repo, IDocu
...
-        _uncached_lookups.fetch_add(1);
-        _store->read(lid, value);
+        if (value.empty()) {
+            return std::unique_ptr<document::Document>();
+        }
+        Value::Result result = value.decompressed();
+        if ( result.second ) {
+            return deserializeDocument(result.first, repo);
+        } else {
+            LOG(warning, "Summary cache for lid %u is corrupt. Invalidating and reading directly backing store", lid);
```
Comment (geirst) [inline searchlib/src/vespa/searchlib/docstore/documentstore.cpp]: Missing word? "directly backing store" -> "directly from backing store"?

**Review by geirst — APPROVED** [review]
Overall looks good, some nitpicks to look at.

**Review by geirst — CHANGES_REQUESTED** [review]
Unit test must be fixed: "384 - searchlib_logdatastore_test_app (Failed)"

**Comment by baldersheim** [comment]
@geirst PR Now it should be OK.

**Review by geirst — COMMENTED** [review]
👍

**Pull request merged**